### PR TITLE
Add functionality for serializing depot manifests

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
This adds functionality for serializing depot manifests into Steam-compatible protobuf files. The output matches .manifest files found in Steam's depotcache folder unless there are non-ASCII filenames present. It seems that Steam handles filename sorting slightly differently, not quite sure what's going on there.

I'm intending to use this as a stepping stone towards migrating DepotDownloader to using Steam protobuf manifests.

### CODE REVIEW

**Review by psychonic — APPROVED** [review]

**Comment by NicknineTheEagle** [comment]
@psychonic How soon is the next SteamKit release? Want to know when I can integrate these changes into DepotDownloader.

**Comment by psychonic** [comment]
I imagine not until the new Steam authentication is worked out and added. (https://github.com/SteamRE/SteamKit/issues/1125 / https://github.com/SteamRE/SteamKit/pull/1129), but I don't really manage the releases so can't say for sure.

**Comment by yaakov-h** [comment]
Usually its whenever I decide to publish one 😄

**Comment by skylayer** [comment]
Thank you for your contribution to the serialization logic of `DepotManifest`. I appreciate the effort you've put into implementing this feature. However, I've identified a critical performance issue within your implementation, specifically related to how unique chunks are determined:

```csharp
if ( !uniqueChunks.Exists( x => x.SequenceEqual( chunk.ChunkID! ) ) )
{
    uniqueChunks.Add( chunk.ChunkID! );
}
```

This approach, unfortunately, leads to each existence check having a time complexity of O(N), resulting in an overall complexity of O(N^2) for this section of the code. This becomes significantly problematic for large manifests, such as those exceeding 100GB, where the serialization process could extend up to 30 minutes or more.

To address this issue and enhance the performance, I suggest utilizing a `HashSet` alongside a custom `IEqualityComparer<byte[]>`. This modification ensures that each check operates in near O(1) time, substantially reducing the total serialization time. Below is a proposed implementation:

```csharp
class ByteArrayComparer : IEqualityComparer<byte[]>
{
    public bool Equals(byte[] x, byte[] y)
    {
        if (ReferenceEquals(x, y)) return true;
        if (x == null || y == null) return false;
        return x.SequenceEqual(y);
    }

    public int GetHashCode(byte[] obj)
    {
        if (obj == null) throw new ArgumentNullException(nameof(obj));
        // Adjust the hash code calculation as needed to ensure uniform distribution
        return BitConverter.ToInt32(obj, 0);
    }
}

// Usage...

var uniqueChunks = new HashSet<byte[]>(new ByteArrayComparer());

foreach (var chunk in file.Chunks)
{
    var protochunk = new ContentManifestPayload.FileMapping.ChunkData()
    {
        sha = chunk.ChunkID,
        crc = BitConverter.ToUInt32(chunk.Checksum!, 0),
        offset = chunk.Offset,
        cb_original = chunk.UncompressedLength,
        cb_compressed = chunk.CompressedLength
    };

    protofile.chunks.Add(protochunk);
    // Directly use HashSet.Add for O(1) complexity
    uniqueChunks.Add(chunk.ChunkID!);
}
```

By implementing this solution, we can significantly improve the performance of the serialization process for large depot manifests. I hope you find this suggestion helpful and consider integrating it into your pull request. Thank you again for your valuable contribution.

**Comment by skylayer** [comment]
Following the discussion on performance improvements, I conducted a comparison test between the old and the new serialization methods for downloading and saving a manifest file. Below are the details and results of the test, with sensitive information redacted.

### Test Code:

```csharp
var download = new ManifestDownloader("fourieran", "<redacted>",
    File.ReadAllText("fourieran_RefreshToken.txt"));

await download.Connect();
var servers = await download.GetServersForSteamPipe();

var mani = await download.DownloadManifestAsync(359320, 359326, 4546458238660215956, servers.First());
Console.WriteLine("Downloaded manifest.");

var timeStart = DateTime.Now;
mani.Manifest.SaveToFile("manifest.txt");
Console.WriteLine($"Saved to manifest.txt. Time: {DateTime.Now - timeStart}");
```

### Old Method Test Results:

  ```
  Connected to Steam! Logging in 'fourieran'...
  Logged on using previous RefreshToken
  License list received
  Downloaded manifest.
  Saved to manifest.txt. Time: 00:10:23.0065890
  ```

### New Method Test Results:

  ```
  Connected to Steam! Logging in 'fourieran'...
  Logged on using previous RefreshToken
  License list received
  Downloaded manifest.
  Saved to manifest.txt. Time: 00:00:00.6132716
  ```

### Comparison:

- The `cmp` command returned nothing, indicating no difference in the content of the manifests downloaded by both methods.
- The new method shows a significant performance improvement, reducing the save time from over 10 minutes to just over 0.6 seconds.

**Comment by skylayer** [comment]
I'm not entirely sure why the original author included the duplicate checks—probably had a good reason. I just worked on speeding up the process they set up. 🤷‍♂️

File: SteamKit2/SteamKit2/Types/DepotManifest.cs [hunk 1]
```diff
@@ -312,6 +370,117 @@ void ParseProtobufManifestMetadata(ContentManifestMetadata metadata)
...
+        {
+            DebugLog.Assert( Files != null, nameof( DepotManifest ), "Files was null when attempting to serialize manifest." );
+
+            var payload = new ContentManifestPayload();
+            var uniqueChunks = new List<byte[]>();
+
+            foreach ( var file in Files )
+            {
+                var protofile = new ContentManifestPayload.FileMapping();
+                protofile.filename = file.FileName.Replace( '/', '\\' );
```
Comment (xPaw) [inline SteamKit2/SteamKit2/Types/DepotManifest.cs:386]: This was incorrectly replacing slashes if the filename was not decrypted.

Discovered while adding a roundtrip test: https://github.com/SteamRE/SteamKit/commit/7559904d5f5d9c90a5d6ca14fe3e0003b720dbdd

**Pull request merged**